File: quizzes/quizzesV2/quizzes8/02-problem.ts
export {};

type NestedTag = string | NestedTag[];

type Post = { tags: NestedTag[] };

type Posts = Record<string, Post>;

const posts: Posts = {
  post_1: { tags: ["js", ["cli", "json"], "parsing"] },
  post_2: { tags: ["js", ["api", ["cli"]]] },
};

const flattenArrayWithObj = (): string[] => {
  const tags = Object.values(posts).flatMap(
    (post) => post.tags.flat(5) as string[],
  );

  return [...new Set(tags)];
};

const flattenArrayWithLoop = (): string[] => {
  const uniqueTags: string[] = [];

  for (const item in posts) {
    const postTags: NestedTag[] = posts[item].tags;

    const flatten = (args) => {
      for (const tag of args) {
        if (Array.isArray(tag)) {
          flatten(tag);
        } else {
          let exists = false;

          for (const item of uniqueTags) {
            if (tag === item) {
              exists = true;
              break;
            }
          }

          if (!exists) uniqueTags.push(tag);
        }
      }
    };

    flatten(postTags);
  }

  return uniqueTags;
};

console.log(JSON.stringify({ data: flattenArrayWithObj() }, null, 2));
console.log(JSON.stringify({ data: flattenArrayWithLoop() }, null, 2));

File: bins10/02-problem.rb
# frozen_string_literal: true

require "json"

posts = {
  "post_1": { "tags": [ "js", [ "cli", "json" ], "parsing" ] },
  "post_2": { "tags": [ "js", [ "api", [ "cli" ] ] ] }
}

def flatten_with_enum(posts)
  posts.values.flat_map do |post|
    post[:tags].flatten
  end.uniq
end

def flatten_with_loop(posts)
  unique_tags = []

  flatten = ->(arr) {
    arr.each do |item|
      if item.is_a? Array
        flatten.call(item)
      elsif !unique_tags.include?(item)
        unique_tags << item
      end
    end
  }

  posts.each_value { |p| flatten.call(p[:tags]) }

  unique_tags
end

puts JSON.pretty_generate({ data: flatten_with_enum(posts) })
puts JSON.pretty_generate({ data: flatten_with_loop(posts) })
